File: server/base_console.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

# Copyright (c) 2015, Matthew Brennan Jones <matthew.brennan.jones@gmail.com>
# emu_archive is a HTML based front end for video game console emulators
# It uses a MIT style license
# It is hosted at: https://github.com/workhorsy/emu_archive
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import os, sys
import json


class BaseConsole(object):
	def __init__(self, config_path):
		self.config_path = config_path
		self.button_map = {}

		# Load the config
		if os.path.isfile(self.config_path):
			with open(self.config_path, 'r') as f:
				self.button_map = json.loads(f.read())

	def set_button_map(self, button_map):
		self.button_map = button_map
		
		# Save the config
		with open(self.config_path, 'w') as f:
			f.write(json.dumps(self.button_map, sort_keys=True, indent=4, separators=(',', ': ')))

	def get_button_map(self):
		return self.button_map

	def goodJoin(self, path_a, path_b):
		path = path_a + path_b
		path = os.path.abspath(path)
		path = path.replace("\\", "/")
		return path

File: client/helpers/base_console.go
// Copyright (c) 2015, Matthew Brennan Jones <matthew.brennan.jones@gmail.com>
// emulators-online is a HTML based front end for video game console emulators
// It uses the GNU AGPL 3 license
// It is hosted at: https://github.com/workhorsy/emulators-online
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package helpers

import (
	//"fmt"
	"strings"
	"io/ioutil"
	"path/filepath"
	"os"
	//"runtime"
	"log"
	//"os/exec"
	//"encoding/base64"
	"encoding/json"
)


type BaseConsole struct {
	config_path string
	button_map map[string]string
}

func NewBaseConsole(config_path string) *BaseConsole {
	self := &BaseConsole{}
	self.config_path = config_path

	// Load the config
	if IsFile(self.config_path) {
		data, err := ioutil.ReadFile(self.config_path)
		if err != nil {
			log.Fatal(err)
		}
		err = json.Unmarshal(data, &self.button_map)
		if err != nil {
			log.Fatal(err)
		}
	}

	return self
}

func (self *BaseConsole) SetButtonMap(button_map map[string]string) error {
	self.button_map = button_map

	// Open the file
	f, err := os.Create(self.config_path)
	if err != nil {
		return err
	}
	defer f.Close()

	// Convert the button map to json
	jsoned_data, err := json.MarshalIndent(self.button_map, "", "\t")
	if err != nil {
		return err
	}

	// Write the jsoned button map to file
	f.Write([]byte(jsoned_data))

	return nil
}

func (self *BaseConsole) GetButtonMap() map[string]string {
	return self.button_map
}

func (self *BaseConsole) goodJoin(path_a string, path_b string) string {
	path := path_a + path_b
	path, _ = filepath.Abs(path)
	path = strings.Replace(path, "\\", "/", -1)
	return path
}

